### PR TITLE
Switch ffmpeg to hls muxer (from segment) to fix premature stop on non-patched ffmpeg

### PR DESCRIPTION
**Pull request by JustAMan**
**Changes**
* Change ffmpeg muxer used in HLS streaming from `segment` to `hls` (`segment` muxer has a bug patch for which is not yet accepted upstream)
* Add more options to enforce keyframes in full transcoding (it should help having segment lengths to be much closer to requested "segment length")
* Add even more debug logging - used for making this happen, but won't hurt to keep it there IMHO
* Fix the condition on "enable break on non-keyframes" - this might be a potentially breaking change, but its previous state (before my fix) didn't make _any sense_ to me whatsoever.

**Issues**
It should really fix https://github.com/jellyfin/jellyfin/issues/1694 without relying to use a patched ffmpeg

### CODE REVIEW

File: Emby.Server.Implementations/HttpServer/HttpListenerHost.cs [hunk 1]
```diff
@@ -539,6 +539,10 @@ public async Task RequestHandler(IHttpRequest httpReq, string urlString, string
             }
             finally
             {
+                if (httpRes.StatusCode >= 500)
+                {
+                    _logger.LogDebug("Sending HTTP Response 500 in response to {Url}", urlToLog);
+                }
```
Comment (cvium) [inline Emby.Server.Implementations/HttpServer/HttpListenerHost.cs]: Missing newline

File: MediaBrowser.Api/Playback/BaseStreamingService.cs [hunk 2]
```diff
@@ -289,16 +289,20 @@ private async Task AcquireResources(StreamState state, CancellationTokenSource c
...
 
+            Logger.LogDebug("Launched ffmpeg process");
             state.TranscodingJob = transcodingJob;
 
             // Important - don't await the log task or we won't be able to kill ffmpeg when the user stops playback
             _ = new JobLogger(Logger).StartStreamingLog(state, process.StandardError.BaseStream, logStream);
 
             // Wait for the file to exist before proceeeding
-            while (!File.Exists(state.WaitForPath ?? outputPath) && !transcodingJob.HasExited)
+            var waitFor = state.WaitForPath ?? outputPath;
```
Comment (cvium) [inline MediaBrowser.Api/Playback/BaseStreamingService.cs]: Not a very descriptive variable name. What is it a path to?

File: Emby.Server.Implementations/HttpServer/HttpListenerHost.cs (hunk 1 above)
Comment (JustAMan) [reply Emby.Server.Implementations/HttpServer/HttpListenerHost.cs]: added

File: MediaBrowser.Api/Playback/BaseStreamingService.cs (hunk 2 above)
Comment (JustAMan) [reply MediaBrowser.Api/Playback/BaseStreamingService.cs]: renamed to `ffmpegTargetFile`

File: MediaBrowser.Api/Playback/BaseStreamingService.cs [hunk 3]
```diff
@@ -289,16 +289,20 @@ private async Task AcquireResources(StreamState state, CancellationTokenSource c
...
 
             // Wait for the file to exist before proceeeding
-            while (!File.Exists(state.WaitForPath ?? outputPath) && !transcodingJob.HasExited)
+            var ffmpegTargetFile = state.WaitForPath ?? outputPath;
+            Logger.LogDebug("Waiting for the creation of {0}", ffmpegTargetFile);
+            while (!File.Exists(ffmpegTargetFile) && !transcodingJob.HasExited)
             {
                 await Task.Delay(100, cancellationTokenSource.Token).ConfigureAwait(false);
             }
+            Logger.LogDebug("File {0} created or transcoding has finished", ffmpegTargetFile);
```
Comment (Bond-009) [inline MediaBrowser.Api/Playback/BaseStreamingService.cs]:
```suggestion

            Logger.LogDebug("File {0} created or transcoding has finished", ffmpegTargetFile);
```

File: MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs [hunk 4]
```diff
@@ -910,8 +925,20 @@ protected override string GetVideoArguments(StreamState state, EncodingOptions e
             {
                 var keyFrameArg = string.Format(
                     " -force_key_frames:0 \"expr:gte(t,{0}+n_forced*{1})\"",
-                    GetStartNumber(state) * state.SegmentLength,
+                    (GetStartNumber(state) * state.SegmentLength).ToString(CultureInfo.InvariantCulture),
```
Comment (Bond-009) [inline MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs]: Don't need to call `ToString`, but the `CultureInfo` should be passed to `string.Format`

File: MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs [hunk 5]
```diff
@@ -910,8 +925,20 @@ protected override string GetVideoArguments(StreamState state, EncodingOptions e
...
+                if (state.TargetFramerate.HasValue)
+                {
+                    // This is to make sure keyframe interval is limited to our segment,
+                    // as forcing keyframes is not enough.
+                    // Example: we encoded half of desired length, then codec detected
+                    // scene cut and inserted a keyframe; next forced keyframe would
+                    // be created outside of segment, which breaks seeking.
+                    keyFrameArg += string.Format(
+                        " -g {0} -keyint_min {0}",
+                        ((int)(state.SegmentLength * state.TargetFramerate)).ToString(CultureInfo.InvariantCulture)
```
Comment (Bond-009) [inline MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs]: Same as above

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs [hunk 6]
```diff
@@ -2168,7 +2168,8 @@ public string GetInputModifier(EncodingJobInfo state, EncodingOptions encodingOp
                 // Important: If this is ever re-enabled, make sure not to use it with wtv because it breaks seeking
                 if (!string.Equals(state.InputContainer, "wtv", StringComparison.OrdinalIgnoreCase)
                     && state.TranscodingType != TranscodingJobType.Progressive
-                    && state.EnableBreakOnNonKeyFrames(outputVideoCodec))
+                    && !state.EnableBreakOnNonKeyFrames(outputVideoCodec) &&
```
Comment (Bond-009) [inline MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs]: `&&` should be at the start of the next line

File: MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs (hunk 4 above)
Comment (JustAMan) [reply MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs]: done

File: MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs (hunk 5 above)
Comment (JustAMan) [reply MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs]: done

File: MediaBrowser.Api/Playback/BaseStreamingService.cs [hunk 7]
```diff
@@ -314,6 +319,7 @@ private async Task AcquireResources(StreamState state, CancellationTokenSource c
             {
                 StartThrottler(state, transcodingJob);
             }
+            Logger.LogDebug("StartFfMpeg() finished successfully");
```
Comment (Bond-009) [inline MediaBrowser.Api/Playback/BaseStreamingService.cs]:
```suggestion
            Logger.LogDebug("StartFfMpeg() finished successfully");

```

File: MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs [hunk 8]
```diff
@@ -458,56 +461,67 @@ private string GetSegmentPath(StreamState state, string playlist, int index)
...
+                    {
+                        segmentExists = File.Exists(segmentPath);
+                        if (segmentExists)
                         {
-                            return await GetSegmentResult(state, segmentPath, segmentIndex, transcodingJob).ConfigureAwait(false);
+                            continue; // avoid unnecessary waiting if segment just became available
                         }
-                        //break;
                     }
+                    await Task.Delay(100, cancellationToken).ConfigureAwait(false);
```
Comment (Bond-009) [inline MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs]:
```suggestion

                    await Task.Delay(100, cancellationToken).ConfigureAwait(false);
```

File: MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs [hunk 9]
```diff
@@ -908,10 +923,22 @@ protected override string GetVideoArguments(StreamState state, EncodingOptions e
             }
             else
             {
-                var keyFrameArg = string.Format(
+                var keyFrameArg = string.Format(CultureInfo.InvariantCulture,
```
Comment (Bond-009) [inline MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs]: All parameters should be on a new line.
```suggestion
                var keyFrameArg = string.Format(
                    CultureInfo.InvariantCulture,
```

File: MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs [hunk 10]
```diff
@@ -908,10 +923,22 @@ protected override string GetVideoArguments(StreamState state, EncodingOptions e
...
-                    state.SegmentLength.ToString(CultureInfo.InvariantCulture));
+                    state.SegmentLength);
+                if (state.TargetFramerate.HasValue)
+                {
+                    // This is to make sure keyframe interval is limited to our segment,
+                    // as forcing keyframes is not enough.
+                    // Example: we encoded half of desired length, then codec detected
+                    // scene cut and inserted a keyframe; next forced keyframe would
+                    // be created outside of segment, which breaks seeking.
+                    keyFrameArg += string.Format(CultureInfo.InvariantCulture,
```
Comment (Bond-009) [inline MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs]: Same
```suggestion
                    keyFrameArg += string.Format(
                        CultureInfo.InvariantCulture,
```

File: MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs [hunk 11]
```diff
@@ -458,56 +461,67 @@ private string GetSegmentPath(StreamState state, string playlist, int index)
...
+                            continue; // avoid unnecessary waiting if segment just became available
                         }
-                        //break;
                     }
+                    await Task.Delay(100, cancellationToken).ConfigureAwait(false);
                 }
-                catch (IOException)
+
+                cancellationToken.ThrowIfCancellationRequested();
+                if (!File.Exists(segmentPath))
```
Comment (Bond-009) [inline MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs]: Switch these around.

Comment (JustAMan) [reply MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs]: Makes sense, fixed. Thanks!

File: MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs [hunk 12]
```diff
@@ -458,56 +461,68 @@ private string GetSegmentPath(StreamState state, string playlist, int index)
...
-                    var text = File.ReadAllText(playlistPath, Encoding.UTF8);
-
-                    // If it appears in the playlist, it's done
-                    if (text.IndexOf(segmentFilename, StringComparison.OrdinalIgnoreCase) != -1)
+                    // To be considered ready, the segment file has to exist AND
+                    // either the transcoding job should be done or next segment should also exist
+                    if (segmentExists)
                     {
-                        if (!segmentFileExists)
+                        if (transcodingJob.HasExited || File.Exists(nextSegmentPath))
```
Comment (dkanada) [inline MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs]: You check for `transcodingJob.HasExited` in the while loop, is it also necessary here?

File: MediaBrowser.Api/Playback/BaseStreamingService.cs [hunk 13]
```diff
@@ -314,6 +319,8 @@ private async Task AcquireResources(StreamState state, CancellationTokenSource c
...
+
```
Comment (dkanada) [inline MediaBrowser.Api/Playback/BaseStreamingService.cs]: extra newline

File: MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs (hunk 12 above)
Comment (JustAMan) [reply MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs]: Yes, it's needed for the correctness, see comment above that if.

**Comment by joshuaboniface** [comment]
Cherry-picked into 10.4.1.

**Pull request merged**